crates/sui-framework/sources/dynamic_field.move

```diff
@@ -0,0 +1,136 @@
+    assert!(option::is_none(&field.value), EFieldAlreadyExists);
+    option::fill(&mut field.value, value);
+}
+
+/// Immutably borrows the `object`s dynamic field with the name specified by `name: Name`.
+/// Aborts with `EFieldDoesNotExist` if the object does not have a field with that name.
+/// Aborts with `EFieldTypeMismatch` if the field exists, but the value does not have the specified
+/// type.
+public fun borrow<Name: copy + drop + store, Value: store>(
+    object: &UID,
+    name: Name,
+): &Value {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    assert!(option::is_some(&field.value), EFieldDoesNotExist);
+    option::borrow(&field.value)
+}
+
+/// Mutably borrows the `object`s dynamic field with the name specified by `name: Name`.
+/// Aborts with `EFieldDoesNotExist` if the object does not have a field with that name.
+/// Aborts with `EFieldTypeMismatch` if the field exists, but the value does not have the specified
+/// type.
+public fun borrow_mut<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): &mut Value {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    assert!(option::is_some(&field.value), EFieldDoesNotExist);
+    option::borrow_mut(&mut field.value)
+}
+
+/// Removes the `object`s dynamic field with the name specified by `name: Name` and returns the
+/// bound value.
+/// Aborts with `EFieldDoesNotExist` if the object does not have a field with that name.
+/// Aborts with `EFieldTypeMismatch` if the field exists, but the value does not have the specified
+/// type.
+public fun remove<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): Value {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    assert!(option::is_some(&field.value), EFieldDoesNotExist);
+    option::extract(&mut field.value)
+}
+
+// TODO implement exists (without the Value type) once we have lamport timestamps
+/// Returns true if and only if the `object` has a dynamic field with the name specified by
+/// `name: Name`.
+public fun exists_with_type<Name: copy + drop + store, Value: store>(
+    object: &UID,
+    name: Name,
+): bool {
+    let object_addr = object::uid_to_address(object);
+    let hash = hash_type_and_key(object_addr, name);
+    if (!has_child_object_with_ty<Field<Name, Value>>(object_addr, hash)) return false;
+    let field = borrow_child_object<Field<Name, Value>>(object_addr, hash);
+    option::is_some(&field.value)
+}
+
+public(friend) native fun hash_type_and_key<K: copy + drop + store>(parent: address, k: K): address;
```

Comment:
One idea for simplification what if this were:

```
fun derive_id_from_type_and_key<T: key, K: copy + drop + store>(parent: T, k: K): ID
```

lived inside `object`, and then `object::new_uid_from_hash` worked by simply wrapping the resulting `ID`? That feels like it might localize the sensitive UID generation logic to `object` a bit more, which seems like a good place for it to live.

Reply:
It's a bit tricky... We almost always need the hash for interfacing with the other native functions (which don't use ID to avoid unnecessary unwrapping). So the only place that actually uses the UID version is `add`, everything else uses the address directly. 

To say this differently, we could make this return a `UID`, but then every function here would call `uid_to_address` immediately, which seems unnecessary